pilot/pkg/model/virtualservice.go

```diff
@@ -35,72 +35,47 @@ import (
 func SelectVirtualServices(vsidx virtualServiceIndex, configNamespace string, hostsByNamespace map[string]hostClassification) []config.Config {
 	importedVirtualServices := make([]config.Config, 0)
 
-	vsset := sets.New[string]()
-	addVirtualService := func(vs config.Config, hosts hostClassification) {
-		vsname := vs.Name + "/" + vs.Namespace
-		rule := vs.Spec.(*networking.VirtualService)
-
-		for _, vh := range rule.Hosts {
-			if vsset.Contains(vsname) {
-				break
-			}
-
-			// first, check exactHosts
-			if hosts.exactHosts.Contains(host.Name(vh)) {
-				importedVirtualServices = append(importedVirtualServices, vs)
-				vsset.Insert(vsname)
-				break
-			}
-
-			// exactHosts not found, fallback to loop allHosts
-			for _, ah := range hosts.allHosts {
-				if vsHostMatches(vh, ah, vs) {
-					importedVirtualServices = append(importedVirtualServices, vs)
-					vsset.Insert(vsname)
-					break
-				}
-			}
-		}
+	addVirtualService := func(vst virtualServiceTrie, host []string) {
+		importedVirtualServices = vst.matchesTrie.Matches(host, importedVirtualServices)
+		importedVirtualServices = vst.subsetOfTrie.SubsetOf(host, importedVirtualServices)
 	}
 
-	loopAndAdd := func(vses []config.Config) {
-		for _, c := range vses {
-			configNamespace := c.Namespace
-			// Selection algorithm:
-			// virtualservices have a list of hosts in the API spec
-			// if any host in the list matches one service hostname, select the virtual service
-			// and break out of the loop.
+	loopAndAdd := func(ct configTrie) {
+		if len(ct.trie) == 0 {
+			return
+		}
 
-			// Check if there is an explicit import of form ns/* or ns/host
-			if importedHosts, nsFound := hostsByNamespace[configNamespace]; nsFound {
-				addVirtualService(c, importedHosts)
+		for crNs, hc := range hostsByNamespace {
+			if crNs == wildcardNamespace {
+				for _, h := range hc.allHosts {
+					for _, vst := range ct.trie {
+						addVirtualService(vst, strings.Split(h.String(), "."))
+					}
+				}
+				continue
 			}
 
-			// Check if there is an import of form */host or */*
-			if importedHosts, wnsFound := hostsByNamespace[wildcardNamespace]; wnsFound {
-				addVirtualService(c, importedHosts)
+			if vst, ok := ct.trie[crNs]; ok {
+				for _, h := range hc.allHosts {
+					addVirtualService(vst, strings.Split(h.String(), "."))
+				}
 			}
 		}
 	}
 
 	n := types.NamespacedName{Namespace: configNamespace, Name: constants.IstioMeshGateway}
+
 	loopAndAdd(vsidx.privateByNamespaceAndGateway[n])
 	loopAndAdd(vsidx.exportedToNamespaceByGateway[n])
 	loopAndAdd(vsidx.publicByGateway[constants.IstioMeshGateway])
 
-	return importedVirtualServices
-}
+	exists := sets.NewWithLength[string](len(importedVirtualServices))
+	slices.FilterInPlace(importedVirtualServices, func(vs config.Config) bool {
+		name := vs.Name + "/" + vs.Namespace
+		return !exists.InsertContains(name)
```

Comment:
As noted in ohter PR, we should not be mutating state inside closures passed to `slices`; they are intended to be idempotent and stateless functions